### Handling imports

In [25]:
import cifar10
from sklearn import datasets
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report,confusion_matrix
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score
import numpy as np

### Next 4 were given in sample template

In [2]:
cifar10.data_path = "data/CIFAR-10/"

In [3]:
cifar10.maybe_download_and_extract()

Data has apparently already been downloaded and unpacked.


In [4]:
class_names = cifar10.load_class_names()
class_names

Loading data: data/CIFAR-10/cifar-10-batches-py/batches.meta


['airplane',
 'automobile',
 'bird',
 'cat',
 'deer',
 'dog',
 'frog',
 'horse',
 'ship',
 'truck']

In [5]:
images_train, cls_train, labels_train = cifar10.load_training_data()
images_test, cls_test, labels_test = cifar10.load_test_data()

Loading data: data/CIFAR-10/cifar-10-batches-py/data_batch_1
Loading data: data/CIFAR-10/cifar-10-batches-py/data_batch_2
Loading data: data/CIFAR-10/cifar-10-batches-py/data_batch_3
Loading data: data/CIFAR-10/cifar-10-batches-py/data_batch_4
Loading data: data/CIFAR-10/cifar-10-batches-py/data_batch_5
Loading data: data/CIFAR-10/cifar-10-batches-py/test_batch


### Looing at shape of training data

In [37]:
cls_train.shape

(50000,)

### images is of shape 50000 ,32,32,3 means 50000 image hai 32*32 ki and 3 is rgb

In [7]:
images_train.shape

(50000, 32, 32, 3)

In [8]:
labels_train.shape

(50000, 10)

### Since images are in mutlidimensiol array we need to convert it into 2d so 50000,32*32*3

In [9]:
x_train = images_train.reshape((50000, 3072))
x_test  = images_test.reshape((10000, 3072))

In [10]:
pca = PCA()
pca.fit(x_train)

PCA()

### Reducing features by keeping 95% variance 

In [11]:
k=0
total=0
while total<0.95:
    total += pca.explained_variance_ratio_[k]
    k += 1
k

217

### Now we know k should be 217 for 95% variance , We will again create pca object with n_components= 217

In [18]:
pca = PCA(n_components=k,whiten=True)
x_transformed = pca.fit_transform(x_train)
x_transformed

array([[-0.86027628,  0.58956777,  0.42062095, ...,  0.62929502,
         0.7761025 ,  0.76742428],
       [ 0.11152023, -0.20522094,  1.68161846, ...,  1.94699703,
        -0.18685917,  0.96716032],
       [ 1.03891411, -2.48917687, -0.77127413, ..., -0.7918517 ,
         0.10204666, -0.55298249],
       ...,
       [ 0.09506512, -2.45038108, -0.11033342, ..., -0.0204035 ,
         1.1420846 , -0.12777842],
       [ 1.52781059, -0.72932918,  1.5590617 , ..., -0.98022143,
         1.53383583, -0.99255869],
       [ 0.56349842, -0.25582611, -1.24529006, ..., -0.25786515,
        -0.23659761, -0.55400581]])

### Getting x_train and y_train

In [13]:
x_train_pca = x_transformed
x_test_pca = pca.transform(x_test)

### Creating object of Random Forest 
prediction is worked by not on running this

In [17]:
clf = RandomForestClassifier()
cross_val_score(clf,x_train_pca,cls_train)
clf.fit(x_train_pca, cls_train)
y_pred = clf.predict(x_test_pca)
print(classification_report(cls_test, y_pred))

              precision    recall  f1-score   support

           0       0.51      0.54      0.53      1000
           1       0.49      0.58      0.53      1000
           2       0.34      0.29      0.32      1000
           3       0.31      0.28      0.29      1000
           4       0.41      0.40      0.41      1000
           5       0.38      0.36      0.37      1000
           6       0.48      0.52      0.50      1000
           7       0.50      0.40      0.45      1000
           8       0.53      0.60      0.56      1000
           9       0.46      0.47      0.47      1000

    accuracy                           0.45     10000
   macro avg       0.44      0.45      0.44     10000
weighted avg       0.44      0.45      0.44     10000



### Cls_test contains numbers like 3 ,2,1,8... and each index specifies a class like 
airplane 0 <br>
 'automobile 1 <br>
 'bird' 2 <br>
 'cat' 3 and son on.... <br>

In [27]:
ans=[]
for i in cls_test :
    ans.append(class_names[i])
    
np.array(ans)

array(['cat', 'ship', 'ship', ..., 'dog', 'automobile', 'horse'],
      dtype='<U10')

### Save the output in csv file

In [35]:
np.savetxt("cifar10_pred_sarthak.csv",np.array(ans),fmt='%s')